filesystem/block.go

```diff
@@ -0,0 +1,340 @@
+			inodeID, addRange.last, count))
+		count++
+	}
+
+	if len(paramStrings) == 0 {
+		// We had only one block, and it was empty. Nothing do to.
+		return nil
+	}
+
+	insStmt := fmt.Sprintf(`INSERT INTO fs.block VALUES %s`, strings.Join(paramStrings, ","))
+	if _, err := e.Exec(insStmt, params...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// read returns the data [from, to).
+// Requires: to > from and [to, from) is contained in the file.
+func read(e sqlExecutor, inodeID, from, to uint64) ([]byte, error) {
+	readRange := newBlockRange(from, to-from)
+	end := readRange.last
+	if readRange.lastLength == 0 {
+		end--
+	}
+
+	blockInfos, err := getBlocksBetween(e, inodeID, readRange.start, end)
+	if err != nil {
+		return nil, err
+	}
+	if len(blockInfos) != end-readRange.start+1 {
+		return nil, fmt.Errorf("wrong number of blocks, asked for [%d-%d], got %d back",
+			readRange.start, end, len(blockInfos))
+	}
+
+	if readRange.lastLength != 0 {
+		// We have a last partial block, truncate it.
+		last := len(blockInfos) - 1
+		blockInfos[last].data = blockInfos[last].data[:readRange.lastLength]
+	}
+	blockInfos[0].data = blockInfos[0].data[readRange.startOffset:]
+
+	var data []byte
+	for _, b := range blockInfos {
+		data = append(data, b.data...)
+	}
+
+	return data, nil
+}
+
+// write commits data to the blocks starting at 'offset'
+// Amount of data to write must be non-zero.
+// If offset is greated than 'originalSize', the file is grown first.
+// We always write all or nothing.
+func write(e sqlExecutor, inodeID, originalSize, offset uint64, data []byte) error {
+	if offset > originalSize {
+		diff := offset - originalSize
+		if diff > BlockSize*2 {
+			// we need to grow the file by at least two blocks. Use growing method
+			// which only sends empty blocks once.
+			if err := grow(e, inodeID, originalSize, offset); err != nil {
+				return err
+			}
+			originalSize = offset
+		} else if diff > 0 {
+			// don't grow the file first, just change what we need to write.
+			data = append(make([]byte, diff, diff), data...)
+			offset = originalSize
+		}
+	}
+
+	// Now we know that offset is <= originalSize.
+	writeRange := newBlockRange(offset, uint64(len(data)))
+	writeFrom := writeRange.start
+
+	if writeRange.startOffset > 0 {
+		// We're partially overwriting a block (this includes appending
+		// to the last block): fetch it, grow it, and update it.
+		// TODO(marc): this would be more efficient if we had RPAD for bytes.
+		blockData, err := getBlockData(e, inodeID, writeRange.start)
+		if err != nil {
+			return err
+		}
+		blockData = append(blockData[:writeRange.startOffset], data[:writeRange.startLength]...)
+		data = data[writeRange.startLength:]
+		if err := updateBlockData(e, inodeID, writeRange.start, blockData); err != nil {
+			return err
+		}
+		// We don't need to insert this block.
+		writeFrom++
+	}
+
+	writeTo := writeRange.last
+	if writeRange.lastLength == 0 {
+		// Last block is empty, don't update/insert it.
```

Comment:
This is purely from the `blockRange` struct. I thought about having it skip the last block when it's empty, but I was already done for the day. I'll probably change it next time I poke at this.